File: 2.Two Pointer/15. 3Sum.py
def threeSum(nums):
    if len(nums) < 3:
        return []
    nums.sort()
    res = set()
    for i in range(len(nums) - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        d = {}
        for j in range(i + 1, len(nums)):
            if nums[j] not in d:
                d[-nums[i] - nums[j]] = j
            else:
                res.add((nums[i], -nums[i] - nums[j], nums[j]))
    return list(res)

File: 2.Two Pointer/15. 3Sum/15.3sum_.go
package main 


func main(){

}


func threeSum(nums []int) [][]int {
    if len(nums) < 3 {
        return [][]int{}
    }
    
    sort.Ints(nums)
    var results [][]int

    for i := 0; i < len(nums)-2; i++ {
        if i > 0 && nums[i] == nums[i-1] {
            continue
        }

        target, left, right := -nums[i], i+1, len(nums)-1

        for left < right {
            sum := nums[left] + nums[right]

            if sum == target {
                results = append(results, []int{-target, nums[left], nums[right]})
                left++
                right--

                for left < right && nums[left] == nums[left-1] {
                    left++
                }
                for left < right && nums[right] == nums[right+1] {
                    right--
                }
            } else if sum < target {
                left++
            } else {
                right--
            }
        }
    }

    return results
}
